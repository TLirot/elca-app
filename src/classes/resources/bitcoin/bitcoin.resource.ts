import { HttpRequest } from "../../network/http-request";
import { BitcoinEndpointEnum } from "./enums/bitcoin-endpoint.enum";
import { CurrentCoinsStatus } from "./interfaces/current-coins-status.interface";

export class BitcoinResource extends HttpRequest {

    async get<T extends { [P in keyof T]: unknown } = CurrentCoinsStatus>(): Promise<T> {
        try{
            const response = await super.get<T>(BitcoinEndpointEnum.currentPrice);
            return response.data;
        }catch (error){
            console.log(error);
            throw error;
        }
    }
}
