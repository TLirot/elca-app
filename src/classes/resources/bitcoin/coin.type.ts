export type CoinType = "USD" | "GBP" | "EUR";
