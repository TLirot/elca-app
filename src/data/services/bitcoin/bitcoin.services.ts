import { HttpRequest } from "../../http-request";
import { CurrentCoinsStatus, ECCurrentCoinsStatus } from "./models/current-coins-status.model";
import { CoindeskApi } from "./coindesk.api";

export class BitcoinServices {

    private httpRequest = new HttpRequest();

    async coinStatus (): Promise<ECCurrentCoinsStatus> {
        try {
            const url = new CoindeskApi().coinsStatus();
            const response = await this.httpRequest.get<CurrentCoinsStatus>(url);
            //TODO mapper
            return response.data as ECCurrentCoinsStatus;
        } catch (error) {
            console.log(error);
            throw error;
        }
    }
}
