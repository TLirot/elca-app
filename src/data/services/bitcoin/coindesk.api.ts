export class CoindeskApi {
    private baseUrl = "https://api.coindesk.com/"

    coinsStatus (): string {
        return this.baseUrl + "v1/bpi/currentprice.json";
    }
}
