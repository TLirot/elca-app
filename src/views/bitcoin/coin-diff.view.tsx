import { Component, CSSProperties, Fragment, ReactNode } from "react";
import { Col } from "react-bootstrap";
import { ArrowBarLeft, ArrowDownCircleFill, ArrowUpCircleFill } from "react-bootstrap-icons";

interface CoinDiffViewProps {
    rate_float: number;
}

interface CoinDiffViewState {
    diff: number;
}

export class CoinDiffView extends Component<CoinDiffViewProps, CoinDiffViewState> {

    readonly defaultDiff = 0;

    readonly state: Readonly<CoinDiffViewState> = {
        diff: this.defaultDiff,
    }

    readonly iconHeight: CSSProperties = { fontSize: "50px" };

    readonly styleArrowUpCircle: CSSProperties = { ...this.iconHeight, color: "green" };

    readonly styleArrowDownCircle: CSSProperties = { ...this.iconHeight, color: "red" };

    get icon (): ReactNode {
        let icon: ReactNode = <ArrowBarLeft style={this.iconHeight}/>;
        if (this.state.diff > this.defaultDiff) {
            icon = <ArrowUpCircleFill style={this.styleArrowUpCircle}/>
        } else if (this.state.diff < 0) {
            icon = <ArrowDownCircleFill style={this.styleArrowDownCircle}/>
        }
        return icon;
    }

    componentDidUpdate (prevProps: Readonly<CoinDiffViewProps>) {
        if (this.props.rate_float !== prevProps.rate_float) {
            this.calculateDiff(prevProps.rate_float);
        }
    }

    calculateDiff (prevRateFloat: number): void {
        const diff = ((prevRateFloat - this.props.rate_float) / prevRateFloat) * 100;
        this.setState({ diff });
    }

    render () {
        return (
            <Fragment>
                <Col className={"col-auto align-self-center"}>
                    {`${this.state.diff} %`}
                </Col>
                <Col className={"col-auto"}>
                    {this.icon}
                </Col>
            </Fragment>
        );
    }
}
