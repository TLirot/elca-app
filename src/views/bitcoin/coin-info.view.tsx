import { Component, CSSProperties, ReactNode } from "react";
import { ECPriceModel } from "../../data/services/bitcoin/models/price.model";
import { Card } from "../../components/layout/card";
import { ArrowBarLeft, ArrowDownCircleFill, ArrowUpCircleFill } from "react-bootstrap-icons";
import { Col, Row } from "react-bootstrap";
import dayjs from "dayjs";

export interface CoinInfoViewProps extends ECPriceModel {
    updateISOTime: string;
}

export interface CoinInfoViewState {
    diff: number;
}

export class CoinInfoView extends Component<CoinInfoViewProps, CoinInfoViewState> {

    readonly defaultDiff = 0;

    readonly state: CoinInfoViewState = {
        diff: this.defaultDiff,
    }

    readonly iconHeight: CSSProperties = { fontSize: "50px" };

    readonly styleArrowUpCircle: CSSProperties = { ...this.iconHeight, color: "green" };
    readonly styleArrowDownCircle: CSSProperties = { ...this.iconHeight, color: "red" };

    get icon (): ReactNode {
        let icon: ReactNode = <ArrowBarLeft style={this.iconHeight}/>;
        if (this.state.diff > this.defaultDiff) {
            icon = <ArrowUpCircleFill style={this.styleArrowUpCircle}/>
        } else if (this.state.diff < 0) {
            icon = <ArrowDownCircleFill style={this.styleArrowDownCircle}/>
        }
        return icon;
    }

    componentDidUpdate (prevProps: Readonly<CoinInfoViewProps>) {
        if (this.props.rate_float !== prevProps.rate_float) {
            this.calculateDiff(prevProps.rate_float);
        }
    }

    calculateDiff (prevRateFloat: number): void {
        const diff = ((prevRateFloat - this.props.rate_float) / prevRateFloat) * 100;
        this.setState({ diff });
    }

    render (): ReactNode {
        return (
            <Card title={`${this.props.code} / ${this.props.description}`}>
                <Row className={"justify-content-between"}>
                    <Col className={"col-auto"}>
                        <img
                            src={"https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcT2WR24fnzSsiHf1TmpIWQn_E3qgJTLBcsK5w&usqp=CAU"}
                            alt="bitcoin"
                            style={{ height: "50px", width: "250px" }}
                        />
                    </Col>
                    <Col className={"col-auto align-self-center"}>
                        {`${this.props.rate} â‚¬`}
                    </Col>
                </Row>
                <Row className={"justify-content-end"}>
                    <Col className={"col-auto align-self-center"}>
                        {`${this.state.diff} %`}
                    </Col>
                    <Col className={"col-auto"}>
                        {this.icon}
                    </Col>
                </Row>
                <hr/>
                <Row>
                    <Col className={"col-auto"}>
                        {dayjs(this.props.updateISOTime).format("MMM-DD HH:mm")}
                    </Col>
                </Row>
            </Card>
        )
    }
}
