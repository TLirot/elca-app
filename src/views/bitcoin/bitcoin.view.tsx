import { Component, ReactNode } from "react";
import { ECCurrentCoinsStatus } from "../../data/services/bitcoin/models/current-coins-status.model";
import { BitcoinServices } from "../../data/services/bitcoin/bitcoin.services";
import { Button, Spinner } from "react-bootstrap";
import { CoinInfoView } from "./coin-info.view";

export interface BitcoinViewState {
    currentCoinsStatus: ECCurrentCoinsStatus | undefined;
    dataError: string | undefined;
}

export class BitcoinView extends Component<Record<never, never>, BitcoinViewState> {

    readonly state: Readonly<BitcoinViewState> = {
        currentCoinsStatus: undefined,
        dataError: undefined,
    }

    constructor (props: Record<never, never>) {
        super(props);

        this.reloadData = this.reloadData.bind(this);
        this.fetchCurrentStatus = this.fetchCurrentStatus.bind(this);
    }

    async componentDidMount (): Promise<void> {
        await this.fetchCurrentStatus();
    }

    async fetchCurrentStatus (): Promise<void> {
        try {
            const result = await new BitcoinServices().coinStatus();
            this.setState({ currentCoinsStatus: result });
        } catch (error) {
            console.error(error);
            this.setState({ dataError: "Network error. Please wait a few minutes and reload the page." });
        }
    }

    reloadData (): void {
        this.setState({ currentCoinsStatus: undefined }, () => {
            setTimeout(this.fetchCurrentStatus, 200);
        });
    }

    render (): ReactNode {
        return (
            this.state.currentCoinsStatus
                ?
                (
                    <div className={"row"}>
                        <div className={"col-4"}>
                            <CoinInfoView
                                {...this.state.currentCoinsStatus.bpi.EUR}
                                updateISOTime={this.state.currentCoinsStatus.time.updatedISO}/>
                        </div>
                        <div className={"col-4"}>
                            <CoinInfoView
                                {...this.state.currentCoinsStatus.bpi.USD}
                                updateISOTime={this.state.currentCoinsStatus.time.updatedISO}/>
                        </div>
                        <div className={"col-4"}>
                            <CoinInfoView
                                {...this.state.currentCoinsStatus.bpi.GBP}
                                updateISOTime={this.state.currentCoinsStatus.time.updatedISO}/>
                        </div>
                        <div className={"row justify-content-around my-3"}>
                            <div className={"col-2"}>
                                <Button id={"reload-button"} onClick={this.reloadData}>Reload</Button>
                            </div>
                        </div>
                    </div>
                )
                :
                this.state.dataError
                    ?
                    this.state.dataError
                    :
                    (
                        <Spinner animation={"grow"}/>
                    )
        );
    }
}
