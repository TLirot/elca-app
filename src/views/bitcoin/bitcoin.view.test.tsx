import { shallow, ShallowWrapper } from "enzyme";
import { BitcoinView, BitcoinViewState } from "./bitcoin.view";
import { ECCurrentCoinsStatus } from "../../data/services/bitcoin/models/current-coins-status.model";
import { Button } from "react-bootstrap";
import { BitcoinServices } from "../../data/services/bitcoin/bitcoin.services";

const mockECCurrentCoinsStatus: ECCurrentCoinsStatus = {
    time: {
        updatedISO: "2021-08-07T12:45:00+00:00"
    },
    bpi: {
        GBP: {
            rate_float: 1,
            rate: "1",
            description: "GBPDescription",
            code: "GBP",
        },
        EUR: {
            rate_float: 1,
            rate: "1",
            description: "EURDescription",
            code: "EUR",
        },
        USD: {
            rate_float: 1,
            rate: "1",
            description: "USDDescription",
            code: "USD",
        }
    }
}

let initialState: BitcoinViewState;

const initShallowComponent = (): ShallowWrapper<Record<never, never>, BitcoinViewState, BitcoinView> => {
    const wrapper = shallow<BitcoinView>(<BitcoinView/>);

    initialState = wrapper.state();

    return wrapper;
};

const mockDidMount = () => {
    jest.spyOn(BitcoinView.prototype, "componentDidMount").mockImplementation();
};
const mockRender = () => {
    jest.spyOn(BitcoinView.prototype, "render").mockImplementation();
};

describe("BitcoinView =>", () => {
    beforeEach(() => {
        jest.restoreAllMocks();
    });

    describe("render =>", () => {
        beforeEach(() => {
            mockDidMount();
        })

        it("Should render correctly when data is loaded", () => {
            const wrapper = initShallowComponent();

            wrapper.setState({ currentCoinsStatus: mockECCurrentCoinsStatus });

            expect(wrapper.find("#reload-button").find(Button).props().onClick).toBe(wrapper.instance().fetchCurrentStatus);
            expect(wrapper).toMatchSnapshot();
        });

        it("Should show the error message when network request has failed", () => {
            const wrapper = initShallowComponent();

            wrapper.setState({ dataError: "network error" });

            expect(wrapper).toMatchSnapshot();
        });

        it("Should show the Spinner when there is no data and there is no error", () => {
            const wrapper = initShallowComponent();

            wrapper.setState({ dataError: "", currentCoinsStatus: undefined });

            expect(wrapper).toMatchSnapshot();
        });
    });

    it("Should the initial state currentStatus undefined and dataError undefined", () => {
        mockDidMount();
        mockRender();
        const wrapper = initShallowComponent();

        expect(wrapper.state()).toStrictEqual({
            currentCoinsStatus: undefined,
            dataError: undefined,
        });
    });

    describe("Methods =>", () => {
        beforeEach(() => {
            mockDidMount();
            mockRender();
        });

        describe("componentDidMount =>", () => {
            it("Should request the data calling fetchCurrentStatus", () => {
                jest.spyOn(BitcoinView.prototype, "componentDidMount").mockRestore();
                const fetchCurrentStatus = jest.spyOn(BitcoinView.prototype, "fetchCurrentStatus").mockImplementation();

                initShallowComponent();

                expect(fetchCurrentStatus).toHaveBeenCalledTimes(1);
                expect(fetchCurrentStatus).toHaveBeenCalledWith();
            });
        });

        describe("fetchCurrentStatus =>", () => {
            it("Should request coinStatus from BitcoinServices and save the response in the state", async () => {
                const response: ECCurrentCoinsStatus = {
                    time: {
                        updatedISO: "my_time",
                    },
                } as ECCurrentCoinsStatus;

                const coinStatus = jest.spyOn(BitcoinServices.prototype, "coinStatus").mockResolvedValue(response);

                const wrapper = initShallowComponent();
                await wrapper.instance().fetchCurrentStatus();

                expect(coinStatus).toHaveBeenCalledTimes(1);
                expect(coinStatus).toHaveBeenCalledWith();

                expect(wrapper.state()).toStrictEqual({
                    ...initialState,
                    currentCoinsStatus: response,
                });
            });

            it("Should set error with 'Network error. Please wait a few minutes and reload the page.' " +
                "in state if there is a network error", async () => {
                jest.spyOn(BitcoinServices.prototype, "coinStatus").mockRejectedValue("error");
                jest.spyOn(console, "error").mockImplementation();

                const wrapper = initShallowComponent();
                await wrapper.instance().fetchCurrentStatus();

                expect(wrapper.state()).toStrictEqual({
                    ...initialState,
                    dataError: "Network error. Please wait a few minutes and reload the page."
                });
            });
        });
    });
})
