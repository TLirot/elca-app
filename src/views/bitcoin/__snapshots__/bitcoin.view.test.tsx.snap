// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitcoinView => render => Should render correctly when data is loaded 1`] = `
<div
  className="row"
>
  <div
    className="col-4"
  >
    <CoinInfoView
      code="EUR"
      description="EURDescription"
      rate="1"
      rate_float={1}
      updateISOTime="2021-08-07T12:45:00+00:00"
    />
  </div>
  <div
    className="col-4"
  >
    <CoinInfoView
      code="USD"
      description="USDDescription"
      rate="1"
      rate_float={1}
      updateISOTime="2021-08-07T12:45:00+00:00"
    />
  </div>
  <div
    className="col-4"
  >
    <CoinInfoView
      code="GBP"
      description="GBPDescription"
      rate="1"
      rate_float={1}
      updateISOTime="2021-08-07T12:45:00+00:00"
    />
  </div>
  <div
    className="row justify-content-around my-3"
  >
    <div
      className="col-2"
    >
      <Button
        active={false}
        disabled={false}
        id="reload-button"
        onClick={[Function]}
        variant="primary"
      >
        Reload
      </Button>
    </div>
  </div>
</div>
`;

exports[`BitcoinView => render => Should show the Spinner when there is no data and there is no error 1`] = `
<Spinner
  animation="grow"
/>
`;

exports[`BitcoinView => render => Should show the error message when network request has failed 1`] = `"network error"`;
